### PR TITLE
Initial faiss interface support

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
This PR adds initial support through the `knn_vector` field type for faiss. This PR allows users to use HNSW method with flat encoding:
```
{
  "target-index" : {
    "mappings" : {
      "properties" : {
        "target-field" : {
          "type" : "knn_vector",
          "dimension" : 2,
          "method" : {
            "engine" : "faiss",
            "space_type" : "l2",
            "name" : "hnsw",
            "parameters" : {
              "m": 48,
              "encoder": {
                "name": "flat"
              }
            }
          }
        }
      }
    }
  }
}

```

In the future, we will add more methods. 

The main changes include adding a MethodComponentContext parameter type, so that components such as encoders, may be configured through parameters, logic to build custom map of parameters including index description to be passed to faiss library, mapper interface changes, etc.

Additionally, added/refactored tests to better cover initial functionality.
 
### Issues Resolved
#27 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/common/KNNConstants.java [hunk 1]
```diff
@@ -26,34 +26,38 @@
...
+    public static final String HNSW_ALGO_EF_CONSTRUCTION = "efConstruction";
+    public static final String HNSW_ALGO_EF_SEARCH = "efSearch";
+    public static final String HNSW_ALGO_INDEX_THREAD_QTY = "indexThreadQty";
+
+    // Faiss specific constants
+    public static final String FAISS_NAME = "faiss";
+    public final static String FAISS_EXTENSION = ".faiss";
+    public static final String INDEX_DESCRIPTION_PARAMETER = "index_description";
+    public static final String METHOD_ENCODER_PARAMETER = "encoder";
+    public static final String METHOD_PARAMETER_NPROBES = "nprobes";
```
Comment (vamshin) [inline src/main/java/org/opensearch/knn/common/KNNConstants.java:59]: Minor: Should we append library name to these constants to differentiate from future libraries. Something like ```FAISS_METHOD_ENCODER_PARAMETER```?

**Review by vamshin — APPROVED** [review]
Minor comment. LGTM! Thanks

File: src/main/java/org/opensearch/knn/common/KNNConstants.java (hunk 1 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/common/KNNConstants.java:59]: Im thinking about refactoring the constant class a little bit. While encoder is only relevant to faiss now, another library may support it. So, my thinking is that I keep it generic. What do you think? 

In a future PR, I was planning on re-organizing the constants.

Comment (vamshin) [reply src/main/java/org/opensearch/knn/common/KNNConstants.java:59]: If there is plan to reorganize these constants. I am good. Thanks

**Review by VijayanB — APPROVED** [review]
LGTM.

**Pull request merged**